### PR TITLE
Streamline Tuple and DataSource, revise _prev handling.

### PR DESCRIPTION
**Pull request by jheer**
Two major changes:
- Simplify the Tuple class for smaller, more efficient code.
- Make Tuple `_prev` object inherit from the current tuple. This will allow the `_prev` object to act as a proper tuple in its own right, simplifying downstream management (e.g., by aggregators).

### CODE REVIEW

**Comment by jheer** [comment]
Open question: does adding the `Object.create` call for `_prev` inheritance cause any of our benchmarks to suffer? (It should actually _help_ with aggregation, but not sure about other cases.)

**Comment by arvind** [comment]
Unfortunately, it does. Object.create appears to be very expensive (perhaps due to wiring up the prototype chains?). Early versions of Tuple had _prev as a full fledged tuple via Object.create, as in this PR, and it was one of the things I had to undo after benchmarking. 

Except for the Aggregator, I can't think of any other operators that require _prev to function as a tuple unto itself. This is why tuple-spoofing was isolated within the Aggregator.

**Comment by arvind** [comment]
Though, actually when I ripped this functionality out of Tuple, none of the operators (including the Aggregator of the time) needed prev to be a full tuple. In fact, now that I think about it, by isolating prev tuple spoofing within Aggregator, we actually run the risk of performing _more_ Object.create calls because prev tuple instances aren't shared anymore. Perhaps we should run our benchmarks again with this PR. 

One idea, if it does prove to be expensive, might be to have Aggregators "promote" _prev to full Tuples.

**Comment by jheer** [comment]
In a couple of experiments I am indeed seeing noticeably larger interactive delays and the profiler attributes to them to Object.create calls. So, we might want to look into your "promotion" idea as an alternative.

**Comment by arvind** [comment]
One thing to keep in mind: currently, _prev is cleared out at the end of
every propagation cycle. This avoids onerous timestamp checks within
operators.

As a result, naive "promotion" (promoting _prev in situ) would continue
to result in poor performance. The Aggregator got around this situation by
stashing prev tuples. Object.create was called only once (per aggregator)
per tuple.

(Side note: if we were to revert to the tuple spoofing within Aggregators,
we should ensure that Object.create is only called once per tuple rather
than per aggr per tuple).

On Tuesday, August 18, 2015, Jeffrey Heer notifications@github.com wrote:

> In a couple of experiments I am indeed seeing noticeably larger
> interactive delays and the profiler attributes to them to Object.create
> calls. So, we might want to look into your "promotion" idea as an
> alternative.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/vega/vega-dataflow/pull/3#issuecomment-132322149.

## 

## -Arvind

Arvind Satyanarayan http://arvindsatya.com/

File: src/DataSource.js [hunk 1]
```diff
@@ -119,27 +92,92 @@ prototype.fire = function(input) {
...
+
+  var graph = this._graph,
+      mutates = 0,
+      collector = this._inputNode,
+      i, node, router, collects;
+
+  for (i=0; i<pipeline.length; ++i) {
+    node = pipeline[i];
+
+    if (!node._collector && node.batch()) {
```
Comment (arvind) [inline src/DataSource.js:108]: My vote would be to move this logic back to `Graph.connect`. That way vega-dataflow, as a standalone library, will correctly honor batch guarantees for all appropriately annotated Node operators, regardless of whether they're part of a DataSource or not.

File: src/DataSource.js [hunk 2]
```diff
@@ -119,27 +92,92 @@ prototype.fire = function(input) {
...
-    ds._collector = new Collector(this._graph);
-    pipeline.push(ds._collector);
-    ds._revises = pipeline.some(function(p) { return p.revises(); });
-    ds._mutates = pipeline.some(function(p) { return p.mutates(); });
+  if (data.length && data[0]._prev) {
+    for (i=0, n=data.length; i<n; ++i) {
+      Tuple.prev_update(data[i]);
+    }
+  }
+  if (this._collector) {
```
Comment (arvind) [inline src/DataSource.js]: Still thinking through the implications of the new `synchronize` method. But, some room for further optimization here if the tuples in `this._collector` are the same ones in `this._data`?

File: src/DataSource.js (hunk 1 above)
Comment (jheer) [inline src/DataSource.js:108]: Fine with me, though doesn't seem very high priority. Could be done pre- or post-merge.

File: src/DataSource.js (hunk 2 above)
Comment (jheer) [inline src/DataSource.js]: Yes, if the tuples are the same then we could cull redundancy here. Will patch...

**Pull request merged**